pallets/nodle-staking/src/lib.rs

```diff
@@ -0,0 +1,2311 @@
+                id,
+                bond,
+                nomi_bond_total: Zero::zero(),
+                nominators: OrderedSet::new(),
+                total,
+                state: ValidatorStatus::default(), // default active
+                unlocking: Vec::new(),
+            }
+        }
+        pub fn is_active(&self) -> bool {
+            self.state == ValidatorStatus::Active
+        }
+        pub fn is_leaving(&self) -> bool {
+            matches!(self.state, ValidatorStatus::Leaving(_))
+        }
+        pub fn bond_more(&mut self, more: B) {
+            self.bond += more;
+            self.total += more;
+        }
+        // Returns None if underflow or less == self.bond (in which case validator should leave)
+        pub fn bond_less(&mut self, less: B) -> Option<B> {
+            if self.bond > less {
+                self.bond -= less;
+                // self.total -= less;
+                Some(self.bond)
+            } else {
+                None
+            }
+        }
+        pub fn inc_nominator(&mut self, nominator: A, more: B) {
+            for x in &mut self.nominators.0 {
+                if x.owner == nominator {
+                    x.amount += more;
+                    self.nomi_bond_total += more;
+                    self.total += more;
+                    return;
+                }
+            }
+        }
+        pub fn dec_nominator(&mut self, nominator: A, less: B) {
+            for x in &mut self.nominators.0 {
+                if x.owner == nominator {
+                    x.amount -= less;
+                    self.nomi_bond_total -= less;
+                    self.total -= less;
+                    return;
+                }
+            }
+        }
+        pub fn go_offline(&mut self) {
+            self.state = ValidatorStatus::Idle;
+        }
+        pub fn go_online(&mut self) {
+            self.state = ValidatorStatus::Active;
+        }
+        pub fn leave_validators_pool(&mut self, round: SessionIndex) {
+            self.state = ValidatorStatus::Leaving(round);
+        }
+    }
+
+    impl<AccountId, Balance> Validator<AccountId, Balance>
+    where
+        Balance: AtLeast32BitUnsigned + Saturating + Copy,
+    {
+        /// Slash the validator for a given amount of balance. This can grow the value
+        /// of the slash in the case that the validator has less than `minimum_balance`
+        /// active funds. Returns the amount of funds actually slashed.
+        ///
+        /// Slashes from `active` funds first, and then `unlocking`, starting with the
+        /// chunks that are closest to unlocking.
+        pub(crate) fn slash(&mut self, mut value: Balance, minimum_balance: Balance) -> Balance {
+            let pre_total = self.total;
+            let total = &mut self.total;
+            let active = &mut self.bond;
+
+            let slash_out_of =
+                |total_remaining: &mut Balance, target: &mut Balance, value: &Balance| {
```

Comment:
I am wondering if it would make sense to turn this in its own function and try to avoid all those pointer things if possible

Reply:
True Eliott, There is subtile variation between the two

slash_out_of in Validator::slash()

works on params

![](https://i.imgur.com/Kxln0JN.png)

slash_out_of in Nominator::slash_nomination()

works on params

![](https://i.imgur.com/0vaJuoF.png)

Both have different control flow.

Right now better to un-touch this optimization :smile: